src/matrixrtc/MatrixRTCSession.ts

```diff
@@ -0,0 +1,385 @@
+/*
+Copyright 2023 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { logger } from "../logger";
+import { TypedEventEmitter } from "../models/typed-event-emitter";
+import { EventTimeline } from "../models/event-timeline";
+import { Room } from "../models/room";
+import { MatrixClient } from "../client";
+import { EventType } from "../@types/event";
+import { CallMembership, CallMembershipData } from "./CallMembership";
+import { Focus } from "./focus";
+import { MatrixEvent } from "../matrix";
+
+const MEMBERSHIP_EXPIRY_TIME = 60 * 60 * 1000;
+const MEMBER_EVENT_CHECK_PERIOD = 2 * 60 * 1000; // How often we check to see if we need to re-send our member event
+const CALL_MEMBER_EVENT_RETRY_DELAY_MIN = 3000;
+
+export enum MatrixRTCSessionEvent {
+    // A member joined, left, or updated a proprty of their membership
+    MembershipsChanged = "memberships_changed",
+    // We joined or left the session (our own local idea of whether we are joined, separate from MembershipsChanged)
+    JoinStateChanged = "join_state_changed",
+}
+
+export type MatrixRTCSessionEventHandlerMap = {
+    [MatrixRTCSessionEvent.MembershipsChanged]: (
+        oldMemberships: CallMembership[],
+        newMemberships: CallMembership[],
+    ) => void;
+    [MatrixRTCSessionEvent.JoinStateChanged]: (isJoined: boolean) => void;
+};
+
+/**
+ * A MatrixRTCSession manages the membership & properties of a MatrixRTC session.
+ * This class doesn't deal with media at all, just membership & properties of a session.
+ */
+export class MatrixRTCSession extends TypedEventEmitter<MatrixRTCSessionEvent, MatrixRTCSessionEventHandlerMap> {
+    // How many ms after we joined the call, that our membership should expire, or undefined
+    // if we're not yet joined
+    private relativeExpiry: number | undefined;
```

Comment:
Sorry if the question is dumb, but does not it not mean that it's not a `MatrixRTCSession`, but rather something like a `LocalMatrixRTCSession`? (i.e. "our" part of the participation in a session instead of a `MatrixRTCSession`?)

Reply:
Well, it represents both since it holds information of what other participants are in the session as well as whether we're joined, so I don't think I'd say it was specific to just our participation. That is, if we called it `LocalMatrixRTCSession`, we wouldn't have something else called `MatrixRTCSession` or `RemoteMatrixRTCSession` I think.